contest/scoreboard.py

```diff
@@ -0,0 +1,139 @@
+        self.start_time = start_time
+
+    def add_user(self,user):
+        self.users.append(user)
+
+    def add_problem(self,scoreboard_problem):
+        self.problems.append(scoreboard_problem)
+
+    def get_problem(self,id):
+        for scoreboard_problem in self.problems:
+            if (scoreboard_problem.id == id):
+                return scoreboard_problem
+
+    #sort by solved descending. if same sort by penalty
+    def sort_users_by_penalty(self):
+        self.users = sorted(self.users, key=methodcaller('penalty',self.start_time))
+        self.users = sorted(self.users, key=methodcaller('solved'), reverse=True)
+
+    def sort_users_by_solved_testcases(self):
+        self.users = sorted(self.users, key=methodcaller('testcases_solved'), reverse=True)
+
+#for scoreboard
+class Scoreboard_Problem:
+    def __init__(self,id,pname,total_testcase):
+        self.id = id
+        self.pname = pname
+        self.total_testcase = total_testcase
+        self.pass_user = 0
+
+    def add_pass_user(self):
+        self.pass_user += 1
+
+class User:
+    def __init__(self,username):
+        self.username = username
+        self.problems = []
+
+    def add_problem(self,problem):
+        self.problems.append(problem)
+
+    def get_problem(self,id):
+        for problem in self.problems:
+            if(problem.id == id):
+                return problem
+
+    def solved(self):
+        count = 0
+        for problem in self.problems:
+            if(problem.is_solved() == True):
+                count += 1
+        return count
+
+    def testcases_solved(self):
+        count = 0
+        for problem in self.problems:
+            count += problem.testcases_solved()
+        return count
+
+    def penalty(self,start_time):
+        penalty = 0
+        for problem in self.problems:
+            penalty += problem.penalty(start_time);
+        return penalty
+
+#for each user
+class Problem:
+    def __init__(self,id):
+        self.submissions = []
+        self.id = id
+
+    def is_solved(self):
+        for submission in self.submissions:
+            if submission.is_solved():
+                return True
+        return False
+
+    def testcases_solved(self):
+        testcases_solved = 0
+        for submission in self.submissions:
+            testcases_solved = max(testcases_solved,submission.pass_testcases)
+        return testcases_solved
+
+    def add_submission(self,submission):
+        self.submissions.append(submission)
+
+    def submit_times(self):
+        return self.submissions.__len__()
+
+    def penalty(self,start_time):
+        minute = 60
+        #every not passed submission should add addtional penalty
+        not_pass_penalty_unit = 20
+        addtional_penalty = 0
+        penalty = sys.maxsize
+        for submission in self.submissions:
+            if submission.is_solved():
+                penalty = min(penalty,(submission.submit_time - start_time).total_seconds()/minute)
```

Comment:
I suggest to use a function to count penalty of every single wrong-tried submission.